### PR TITLE
Integration tests for nested mutations support

### PR DESCRIPTION
**Pull request by andrisak**
Added tests for nested mutations and fixed a couple of regressions.

### CODE REVIEW

**Comment by Eldow** [comment]
Thank you very much ! It's definitely a great addition.

**Comment by Eldow** [comment]
On a second look, I noticed a small change that leads to updateBy[UniqueColumn] to not work anymore. I'm pretty sure there is an inflector to detect if a field is an updater field but I need to dig deeper into the plugin code.

File: src/PostgraphileNestedMutationsPlugin.js [hunk 1]
```diff
@@ -641,32 +651,41 @@ module.exports = function PostGraphileNestedMutationPlugin(builder) {
               );
             }
 
-            await Promise.all(Object.keys(fieldValue).map(async k => {
-              if (k.includes('updateBy')) {
-                await Promise.all(
-                  (Array.isArray(fieldValue[k]) ? fieldValue[k] : [fieldValue[k]]).map(async (rowData) => {
-                    const resolver = pgNestedUpdateResolvers[foreignTable.id];
+            if (fieldValue.updateById || fieldValue.updateByNodeId) {
```
Comment (Eldow) [inline src/PostgraphileNestedMutationsPlugin.js:654]: This won't handle `updateBy[UniqueField]` cases

File: src/PostgraphileNestedMutationsPlugin.js [hunk 2]
```diff
@@ -142,6 +141,41 @@ module.exports = function PostGraphileNestedMutationPlugin(builder) {
             } = nestedField;
             const fieldValue = input[fieldName];
 
+            if (fieldValue.updateById || fieldValue.updateByNodeId) {
```
Comment (Eldow) [inline src/PostgraphileNestedMutationsPlugin.js:144]: This won't handle `updateBy[UniqueField]` cases

**Review by Eldow — COMMENTED** [review]
I noticed a small case that is not supported by your implementation, the plugin also adds updateBy[UniqueField] inflectors, depending on the unique keys of your postgresql table.
For example, a user table with a unique email will inflect `updateByEmail`.

I'm not so proud of the way I was handling the case, `k.includes('updateBy')` is definitely not the nicest bit of code I wrote and I'm pretty sure there are helpers built-in to achieve the same result.

File: src/PostgraphileNestedMutationsPlugin.js (hunk 1 above)
Comment (andrisak) [reply src/PostgraphileNestedMutationsPlugin.js:654]: Does this test case cover the case you describe?

```javascript
test(
  'forward nested mutation with nested updateByEmail',
  withSchema({
    setup: `
      create table p.parent (
        email text not null primary key,
        name text not null
      );
      
      create table p.child (
        id2 serial primary key,
        parent_email text not null,
        description text not null,
        constraint child_parent_fkey foreign key (parent_email)
          references p.parent (email)
      );

      insert into p.parent values('test@test.com', 'test parent');
      insert into p.child values(1, 'test@test.com', 'test child');
    `,
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          updateParentByEmail(
            input: {
              email: "test@test.com"
              parentPatch: {
                childrenUsingEmail: {
                  updateById2: {
                    id2: 1
                    childPatch: {
                      description: "renamed child"
                    }
                  }
                }
              }
            }
          ) {
            parent {
              email
              name
              childrenByParentEmail {
                nodes {
                  id2
                  parentEmail
                  description
                }
              }
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');

      const data = result.data.updateParentByEmail.parent;
      expect(data.childrenByParentEmail.nodes).toHaveLength(1);
      data.childrenByParentEmail.nodes.map((n) =>
        expect(n.parentEmail).toBe(data.email),
      );
      expect(data.childrenByParentEmail.nodes[0].description).toEqual(
        'renamed child',
      );
    },
  }),
);
```

Comment (Eldow) [reply src/PostgraphileNestedMutationsPlugin.js:654]: Yes, completely. Also when using postgraphile with the --classicIds option enabled, `updateByNodeId` becomes `updateById` and `updateById` becomes `updateByRowId` (which can be quite confusing, but that's how relay works), that's why I was sticking with the more generic `str.includes('updateBy')`, but I think there is a nicer way of checking this, maybe using `pgNestedTableUpdaterFields`.

I noticed those too : 
`isPgCreateMutationField`
`isPgUpdateMutationField`

I'll give it a try next week !

Comment (andrisak) [reply src/PostgraphileNestedMutationsPlugin.js:654]: That test is passing on master, but maybe there is a cleaner/better way to do the check.

Comment (Eldow) [reply src/PostgraphileNestedMutationsPlugin.js:654]: I misunderstood how it all worked on this specific point, thanks for taking the time to check it out and clear up my doubts! I am closing this discussion as everything is working as expected.

**Comment by francescovenica** [comment]
hello, I'm having some issue updating nested children, in particular on the 3rd level, is this a known issue?

**Pull request merged**